api/src/DataProvider/TopBookDataProvider.php

```diff
@@ -0,0 +1,93 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\DataProvider;
+
+use App\Entity\TopBook;
+use Symfony\Component\HttpKernel\KernelInterface;
+use Symfony\Contracts\Cache\CacheInterface;
+use Symfony\Contracts\Cache\ItemInterface;
+
+final class TopBookDataProvider
+{
+    private const MAX_CACHE_TIME = 3600; // 1 hour
+    private const DATA_SOURCE = 'top-100-novel-sci-fi-fr.csv';
+    private const FIELDS_COUNT = 5;
+
+    private CacheInterface $cache;
+    private KernelInterface $kernel;
+
+    public function __construct(CacheInterface $cache, KernelInterface $kernel)
+    {
+        $this->cache = $cache;
+        $this->kernel = $kernel;
+    }
+
+    /**
+     * Local caching is done so the CSV isn't reloaded at every call.
+     */
+    public function getTopBooks(): array
+    {
+        return $this->cache->get('books.sci-fi.top.fr', function (ItemInterface $item) {
+            $item->expiresAfter(self::MAX_CACHE_TIME);
+
+            return $this->getTopBooksFromCsv();
+        });
+    }
+
+    /**
+     * Be careful that the file is a simple csv file without "enclosure". That means
+     * a field can't contain a ";" or this would add an extra column to the row.
+     * Consider using a more robust library like csv reader from the PHP league.
+     *
+     * @see https://csv.thephpleague.com
+     */
+    public function getTopBooksFromCsv(): array
+    {
+        foreach ($this->getFileAsArray() as $line) {
+            $data[] = str_getcsv($line, ';');
+        }
+
+        $cpt = 0;
+        foreach ($data ?? [] as $row) {
+            if (1 === ++$cpt) {
+                continue;
+            }
+            if (self::FIELDS_COUNT !== count($row)) {
+                throw new \RuntimeException(sprintf('Invalid data at row: %d', count($row)));
+            }
+            $topBooks[$cpt - 1] = (new TopBook())
```

Comment:
Quality tools may throw a warning/an error as this variable is not set

Reply:
It's `$data ?? []` not `$data`.

Reply:
There is a try/catch so it's OK I think.